Adding new features to test to the shape_main application

angelrti opened the pull request:
This PR aims to add new features to the interoperability tests. I have modified the shape_main application to add these new options.

I have only added some `#ifdef` blocks with the new options, where in some cases I've just added a printf saying "not supported" because of potential API differences. Please review this PR and try to build this new application with your product.

Feel free to add your changes to this same branch.

Additionally, I will be modifying the `test_suite.py` to add new test scenarios (based on these new options).

ClarkTucker commented on srcCxx/shape_main.cxx:
```diff
@@ -732,12 +959,44 @@ class ShapeApplication {
+            logger.log_message("Error allocating memory for topics", Verbosity::ERROR);
+            return false;
+        }
+        for (int i = 0; i < options->num_topics; ++i) {
```
should probably be "for (unsigned int i = 0; ... "

angelrti replied:
Thank you! My compiler didn't show these warnings. This should be resolved.

ClarkTucker commented:
I've looked at it a little more, and have some more thoughts:

What do you think of changing the command line options a little bit?  Like this:

```
--access_scope  ( i | t | g )    : sets Presentation.access_scope to INSTANCE, TOPIC or GROUP
--coherent                             : sets Presentation.coherent_access = true
--ordered                               : sets Presentation.ordered_access = true
```

And, I think the run_subscriber() code is missing calls to sub->begin_access() and end_access() for the "scope == GROUP" case.

jrw972 commented on srcCxx/shape_main.cxx:
```diff
@@ -732,12 +974,44 @@ class ShapeApplication {
         if (dp)  dp->delete_contained_entities( );
         if (dpf) dpf->delete_participant( dp );
 
+        free(topics);
```
Is there a particular reason you are using malloc/free as opposed to new/delete with STL containers?

angelrti replied:
no specific reasons for using the C API

jrw972 commented:
We built your branch.  There are two errors that should be easy to fix.  Let me know how we can help.
https://github.com/OpenDDS/OpenDDS/actions/runs/13574664165/job/37947859144

angelrti commented:
> We built your branch. There are two errors that should be easy to fix. Let me know how we can help. https://github.com/OpenDDS/OpenDDS/actions/runs/13574664165/job/37947859144

thanks for pointing out, this should be fixed

angelrti commented:
> I've looked at it a little more, and have some more thoughts:
> 
> What do you think of changing the command line options a little bit? Like this:
> 
> ```
> --access_scope  ( i | t | g )    : sets Presentation.access_scope to INSTANCE, TOPIC or GROUP
> --coherent                             : sets Presentation.coherent_access = true
> --ordered                               : sets Presentation.ordered_access = true
> ```
> 
> And, I think the run_subscriber() code is missing calls to sub->begin_access() and end_access() for the "scope == GROUP" case.

that sounds good to me

Thanks for pointing out the begin_access and end_access, I forgot to add it

jrw972 commented:
> > We built your branch. There are two errors that should be easy to fix. Let me know how we can help. https://github.com/OpenDDS/OpenDDS/actions/runs/13574664165/job/37947859144
> 
> thanks for pointing this out, this should be fixed

It looks like the accessor issue is still there https://github.com/OpenDDS/OpenDDS/actions/runs/13574664165/job/38186445678

angelrti commented:
I've added the parameters that Clark mentioned and also --take-read to use take() instead of take_next_instance() (and the same for read)

I've added the new tests scenarios on python. I'm missing to test coherent access and also the documentation of the previous tests.

I won't be available till 13 March, in case that you see that I don't reply, but feel free to add the changes to the C++ code to support your product (I added some `#ifdef` blocks with the new functionality, if not modified, the application prints 'Feature X not supported')

MiguelCompany commented on srcCxx/shape_main.cxx:
```diff
@@ -1061,89 +1533,173 @@ class ShapeApplication {
+                                        sample->shapesize FIELD_ACCESSOR );
+                                if (DDS_UInt8Seq_get_length(&sample->additional_payload_size FIELD_ACCESSOR) > 0) {
+                                    int additional_payload_index = DDS_UInt8Seq_get_length(&sample->additional_payload_size FIELD_ACCESSOR) - 1;
+                                    printf(" {%u}", sample->additional_payload_size[additional_payload_index] FIELD_ACCESSOR);
```
Done in #59

angelrti commented:
> > > We built your branch. There are two errors that should be easy to fix. Let me know how we can help. https://github.com/OpenDDS/OpenDDS/actions/runs/13574664165/job/37947859144
> > 
> > 
> > thanks for pointing this out, this should be fixed
> 
> It looks like the accessor issue is still there https://github.com/OpenDDS/OpenDDS/actions/runs/13574664165/job/38186445678

@jrw972 I think that the issue is the function `DDS_UInt8Seq_get_length`

MiguelCompany commented on srcCxx/shape_main.cxx:
```diff
@@ -1904,6 +1904,15 @@ class ShapeApplication {
             }
         }
 
+#if defined(TWINOAKS_COREDX)
```
Might be correct to do this for all vendors

ClarkTucker replied:
Yes, I agree, but didn't want to presume...

There's also one more change that I didn't include in this commit that I think should be done for everyone:

```
diff --git a/srcCxx/shape_main.cxx b/srcCxx/shape_main.cxx
index 0af6e38..9de51e1 100644
--- a/srcCxx/shape_main.cxx
+++ b/srcCxx/shape_main.cxx
@@ -1063,7 +1063,7 @@ public:
 #endif
         logger.log_message("Running initialize() function", Verbosity::DEBUG);
 
-        DomainParticipantFactory *dpf = OBTAIN_DOMAIN_PARTICIPANT_FACTORY;
+        dpf = OBTAIN_DOMAIN_PARTICIPANT_FACTORY;
         if (dpf == NULL) {
             logger.log_message("failed to create participant factory (missing license?).", Verbosity::ERROR);
             return false;
```

Currently, we are not using the class member "dpf", and so during the destructor we don't ever call delete_contained_entities() and delete_participant().

ClarkTucker replied:
It seems rare for the application to run to completion -- usually, it gets 'killed' by the test driver.  
So, this code really doesn't happen very often...

We currently have a handler for SIGINT that should allow the application to terminate cleanly, but it doesn't seem to be happening, at least I don't think it is...  Do you know what signal is used by the test driver to forcibly terminate the executables?

angelrti replied:
I have also added the `wait_for_acknowledgments` for everybody

GerardoPardo reviewed (approved):
Seems good to me as well

mitza-oci commented:
Even though @jrw972 participated at various times during this PR and contributed code, the resulting merge broke compatibility with OpenDDS (it was due to the "Added wait_for_ack..." commit which is after we reviewed).

Given this code:
```
#if defined(RTI_CONNEXT_DDS)
        Duration_t max_wait = {1, 0}; /* should not take long... */
#else
        Duration_t max_wait( 1, 0 ); /* should not take long... */
#endif
```

OpenDDS works the same way as RTI in this case, and doesn't compile using the other branch.  So a simple fix would be

```
#if defined(RTI_CONNEXT_DDS) || defined (OPENDDS)
        Duration_t max_wait = {1, 0}; /* should not take long... */
```

and I can easily make a PR for this.

What I'd like to know from the group is if we should use GitHub Actions workflows as CI to check for these issues before merging.  There can be any number of CI checks, so anyone could add one for the product they support.
